### PR TITLE
Extracting UcfgBuilder and InstructionVisitor

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Security/Ucfg/UcfgBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            basicBlock.Instructions.Add(instruction);
+
+            MapToVariable(syntaxNode, instruction.Variable);
+        }
+
+        private string CreateTempVariable() =>
+            $"%{tempVariablesCounter++}";
+
+        private static Expression CreateVariableExpression(string name) =>
+            new Expression { Var = new Variable { Name = name } };
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Security/Ucfg/UcfgBuilder.cs]: This line appears twice in this PR, it could probably be refactored.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/InstructionVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Protobuf.Ucfg;
+
+namespace SonarAnalyzer.ControlFlowGraph.CSharp
+{
+    public class InstructionVisitor : CSharpSyntaxVisitor
+    {
+        private readonly SemanticModel semanticModel;
+
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/InstructionVisitor.cs]: Un-needed blank line

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UniversalControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -18,10 +18,8 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UniversalControlFlowGraphBuilder.cs]: Please rename one of the 2 UcfgBuilder classes.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Protobuf.Ucfg;
+
+namespace SonarAnalyzer.ControlFlowGraph.CSharp
+{
+    public class UcfgBuilder
+    {
+        private readonly UcfgObjectFactory objectFactory;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs]: I'll wait to see what this class is doing but `objectFactory` is really vague name...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            }
+            else
+            {
+                ucfg.Entries.Add(blockIdProvider.Get(cfg.EntryBlock));
+            }
+
+            return ucfg;
+        }
+
+        private BasicBlock CreateBasicBlock(Block block)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs]: I thought we said on Friday that we will have`UcfgBuilder`, `UcfgBlockBuilder`, `UcfgInstructionBuilder`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+            foreach (var parameter in methodSymbol.Parameters)
+            {
+                ucfgBlock.Instructions.AddRange(CreateParameterInstructions(parameter));
+            }
+
+            return ucfgBlock;
+        }
+
+        private IEnumerable<Instruction> CreateParameterInstructions(IParameterSymbol parameter) =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs]: I don't think this belongs here... You have a specialized class building instructions.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            parameter.GetAttributes()
+                .Where(a => a.AttributeConstructor != null)
+                .SelectMany(a =>
+                    CreateAttributeInstructions(
+                        a.ApplicationSyntaxReference.GetSyntax(),
+                        UcfgMethod.Create(a.AttributeConstructor),
+                        parameter.Name));
+
+        private IEnumerable<Instruction> CreateAttributeInstructions(SyntaxNode attributeSyntax, UcfgMethod method, string parameterName) =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs]: I don't think this belongs here... You have a specialized class building instructions.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionFactory.cs [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+            }
+            else
+            {
+                return objectFactory.CreateConstant(identifierName);
+            }
+        }
+
+        private Instruction CreateFromVariableDeclarator(VariableDeclaratorSyntax variableDeclarator)
+        {
+            if (variableDeclarator.Initializer != null)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionFactory.cs]: [minor] You could exit early instead of having this nesting.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs [hunk 9]
```diff
@@ -18,29 +18,29 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-using SonarAnalyzer.Protobuf.Ucfg;
+using Microsoft.CodeAnalysis;
 
 namespace SonarAnalyzer.ControlFlowGraph.CSharp
 {
-    internal static class UcfgBuiltIn
+    public struct UcfgMethod
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs]: I would maybe add the `Id` suffix.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Protobuf.Ucfg;
+using UcfgLocation = SonarAnalyzer.Protobuf.Ucfg.Location;
+
+namespace SonarAnalyzer.ControlFlowGraph.CSharp
+{
+    public class UcfgObjectFactory
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: This smells like a kitchen sink class. Let's put everything in here :c

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        private static readonly Expression ConstantExpression = new Expression
+        {
+            Const = new Constant { Value = "\"\"" }
+        };
+
+        private readonly Dictionary<SyntaxNode, Expression> nodeExpressionMap = new Dictionary<SyntaxNode, Expression>();
+
+        private int tempVariablesCounter;
+
+        public UCFG CreateUcfg(SyntaxNode syntaxNode, UcfgMethod method) =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: This belongs to the `UcfgBuilder`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            new UCFG
+            {
+                Location = GetLocation(syntaxNode),
+                MethodId = method.ToString(),
+            };
+
+        public BasicBlock CreateUcfgBlock(string id) =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: This belongs to the `UcfgBlockBuilder`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+            new BasicBlock { Id = id };
+
+        public Instruction CreateMethodCall(SyntaxNode syntaxNode, UcfgMethod method, params SyntaxNode[] arguments) =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: All of the methods returning `Instruction` should belong to the `UcfgInstructionBuilder`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+            {
+                Location = syntaxNode == null
+                    ? null
+                    : GetLocation(syntaxNode),
+                ReturnedExpression = returnedValue == null
+                    ? ConstantExpression
+                    : GetMappedExpression(returnedValue),
+            };
+
+        public Jump CreateJump(params string[] blockIds)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: As far as I understand this should belong to the `UcfgBuilder` class.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+namespace SonarAnalyzer.ControlFlowGraph.CSharp
+{
+    public class UcfgObjectFactory
+    {
+        /// <summary>
+        /// The string constant representation in the Sonar Security engine (Java part). When
+        /// an instruction receives or returns a type that is not string we use this instead
+        /// of a variable.
+        /// </summary>
+        private static readonly Expression ConstantExpression = new Expression
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: Everything related to expression could belong to an `UcfgExpressionBuilder` class.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs [hunk 16]
```diff
@@ -59,9 +59,9 @@ void Foo(string a, string b)
 0 [shape=record label=""{BLOCK:0|%0 __concat b,a|x __id %0|TERMINATOR JUMP: 1, 2}""]
 0 -> 1
 0 -> 2
-1 [shape=record label=""{BLOCK:1|%0 C.Bar(string, int) a,CONST|TERMINATOR JUMP: 3}""]
+1 [shape=record label=""{BLOCK:1|%1 C.Bar(string, int) a,CONST|TERMINATOR JUMP: 3}""]
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs]: Why this change?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs [hunk 17]
```diff
@@ -59,9 +59,9 @@ void Foo(string a, string b)
...
 1 -> 3
-2 [shape=record label=""{BLOCK:2|%0 C.Bar(string, int) b,CONST|TERMINATOR JUMP: 3}""]
+2 [shape=record label=""{BLOCK:2|%2 C.Bar(string, int) b,CONST|TERMINATOR JUMP: 3}""]
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs]: Same here?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs (hunk 5 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs]: I'll leave splitting UcfgBuilder to UcfgBlockBuilder and UcfgBuilder to you and @duncanp-sonar :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs (hunk 10 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: I don't think separate factory for each type will simplify things. Any ideas how to improve?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs (hunk 14 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: Nope. UcfgBuilder generates UCFG CFG; UcfgObjectFactory generates UCFG objects (the protobuf generated classes) from SyntaxNodes.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs (hunk 16 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs]: The temp variables do not start from %0 for each block, the numbering is shared for the whole UCFG.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs (hunk 17 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs]: The temp variables do not start from %0 for each block, the numbering is shared for the whole UCFG.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs (hunk 16 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs]: Because the UcfgObjectFactory (which contains the temp variable counter) is not created for each block, but shared for all blocks.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs (hunk 14 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: Jump is a block element? Is it some instruction?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs (hunk 16 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs]: It used to start at 0, why this change? Is it not going to "break" something?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs (hunk 13 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: UcfgInstructionFactory is taking care of calling the correct method of the UcfgObjectFactory (which instantiates the Instruction class). The separation is nice imo because one class works with Syntax, other with UCFG objects.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs (hunk 11 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs]: See my other comments about the purpose of UcfgObjectFactory. It could be moved in UcfgBuilder, but you'll have to make the UcfgObjectFactory.GetLocation method public... I'd like it to stay here for now.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs [hunk 18]
```diff
@@ -68,16 +68,38 @@ internal static class UcfgBuiltIn
...
         {
-            Const = new Constant { Value = "\"\"" }
-        };
+            switch (methodSymbol?.MethodKind)
+            {
+                case MethodKind.ExplicitInterfaceImplementation:
+                    return Create(methodSymbol.ConstructedFrom.ToDisplayString());
+
+                case MethodKind.ReducedExtension:
+                    return Create(methodSymbol.ReducedFrom.ToDisplayString());
```
[the remark this answers is not in the document]
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs]: Nope, FP

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs [hunk 19]
```diff
@@ -68,16 +68,38 @@ internal static class UcfgBuiltIn
...
+            method.ToString();
+
+        public static UcfgMethod Create(IMethodSymbol methodSymbol)
         {
-            Const = new Constant { Value = "\"\"" }
-        };
+            switch (methodSymbol?.MethodKind)
+            {
+                case MethodKind.ExplicitInterfaceImplementation:
+                    return Create(methodSymbol.ConstructedFrom.ToDisplayString());
```
[the remark this answers is not in the document]
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs]: Nope, FP

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs (hunk 5 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs]: In general the *Builder classes imply implementing Builder pattern, which we are not. I could change the UcfgBuilder to CfgToUcfgConverter but the name is clumsy...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionFactory.cs (hunk 8 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionFactory.cs]: I'll have to put two `return null` statements if I do that.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs (hunk 5 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs]: I started to disagree when saw that having two classes sharing the same state and having two methods each will not improve a lot. It is not a lot of work to separate, I just think the current class is not that bad to require changes.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs (hunk 16 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/UcfgSerializerTest.cs]: It used to create new InstructionBuilder (the old name) for each block. Now the instruction generation is shared for the whole UCFG. It could be like before, but there is no need.
